### PR TITLE
Fixes #25217 - Update the eslint configuration to use eslint-plugin-patternfly-react

### PR DESCRIPTION
**Pull request by sharvit**
I know this PR is big and i'm sorry about that, this is what happen when you change linting configurations.

This PR is divided into 3 different commits to make the review process easier.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -153,6 +153,7 @@
     ],
     "setupFiles": [
       "raf/polyfill",
+      "jest-prop-type-error",
```
Comment (sharvit) [inline package.json]: This will throw errors in test when we get prop-type warnings

File: .eslintrc [hunk 2]
```diff
@@ -1,15 +1,12 @@
...
-  "env": {
-    "browser": true,
-    "es6": true,
-    "node": true,
-    "jasmine": true,
-    "jest": true
+  "plugins": ["patternfly-react"],
+  "extends": ["plugin:patternfly-react/recommended"],
+  "rules": {
+    "import/extensions": "off", # https://github.com/houndci/hound/issues/1419#issuecomment-328727830
```
Comment (ohadlevy) [inline .eslintrc]: do we still care about hound issues (since we disabled it for JS?)

File: package.json (hunk 1 above)
Comment (ohadlevy) [reply package.json]: maybe we can add it as a separate PR? I'm pretty sure this can be merged very quickly.

File: webpack/assets/javascripts/react_app/common/__mocks__/I18n.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+const cheveronPrefix = () => (window.I18N_MARK ? '\u00BB' : '');
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/__mocks__/I18n.js]: why is this needed?

File: webpack/assets/javascripts/react_app/components/AuditsList/UserDetails.js [hunk 4]
```diff
@@ -25,7 +25,7 @@ const UserDetails = ({ isAuditLogin, userInfo, remoteAddress }) => {
           </span>
         </EllipsisWithTooltip>
         <span>
-          <a href={auditPath}>__('Logged-in')</a>
+          <a href={auditPath}>{__('Logged-in')}</a>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/AuditsList/UserDetails.js]: might make sense to create a separate PR for it assuming we want to cherry pick this.

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ const documentationBlock = ({
   url && (
     <React.Fragment>
       {label}{' '}
-      <a href={url} target="_blank">
+      <a href={url} target="_blank" rel="noopener noreferrer">
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: why?

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/

File: package.json (hunk 1 above)
Comment (sharvit) [reply package.json]: https://github.com/theforeman/foreman/pull/6161

Comment (sharvit) [reply package.json]: Rebased

File: webpack/assets/javascripts/react_app/common/__mocks__/I18n.js (hunk 3 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/common/__mocks__/I18n.js]: We don't actually need the chevron in our tests. Removed.

File: .eslintrc (hunk 2 above)
Comment (sharvit) [reply .eslintrc]: Nice catch! Removed.

File: webpack/assets/javascripts/react_app/common/__mocks__/I18n.js (hunk 3 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/common/__mocks__/I18n.js]: No needed, another PR created for the I18n:
https://github.com/theforeman/foreman/pull/6162

File: webpack/assets/javascripts/react_app/components/AuditsList/UserDetails.js (hunk 4 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AuditsList/UserDetails.js]: https://github.com/theforeman/foreman/pull/6163 created

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AuditsList/UserDetails.js]: Rebase

**Comment by sharvit** [comment]
The reasons it still WIP:
- [x] Waiting https://github.com/theforeman/foreman/pull/6162 to be merged
- [x] I broke the Statistics page, need to fix it

**Comment by sharvit** [comment]
@ohadlevy can you rereview please? I removed the WIP

**Comment by ohadlevy** [comment]
@amirfefer can you review please?

File: package.json [hunk 6]
```diff
@@ -11,8 +11,8 @@
     "@storybook/storybook-deployer": "^2.0.0",
     "axios-mock-adapter": "^1.10.0",
     "babel-cli": "^6.10.1",
-    "babel-core": "^6.26.3",
-    "babel-eslint": "^6.1.2",
+    "babel-core": "^6.26.0",
+    "babel-eslint": "^8.2.3",
```
Comment (ohadlevy) [inline package.json:15]: @sharvit can you create packaging PR's to match? thanks. (or remove unneeded changes?)

File: webpack/assets/javascripts/compute_resource/ec2.js [hunk 7]
```diff
@@ -4,20 +4,19 @@ export function vpcSelected({ value }) {
   const sgSelect = $('select.security_group_ids');
   const securityGroups = JSON.parse(sgSelect.attr('data-security-groups'));
   const subnets = JSON.parse(sgSelect.attr('data-subnets'));
-  // eslint-disable-next-line camelcase
```
Comment (ohadlevy) [inline webpack/assets/javascripts/compute_resource/ec2.js]: probably can create a separate pr for these, will be easy/quick to merge it.

**Comment by ohadlevy** [comment]
@sharvit looking at the 3rd commit, imho everything that can be broken down and merged directly into develop branch now, should probably be one pr, ideally we would end up with only the eslint config change + automatic changes?

File: package.json [hunk 8]
```diff
@@ -43,6 +41,7 @@
     "jsdom": "^9.5.0",
     "lodash-webpack-plugin": "^0.11.4",
     "node-sass": "^4.5.0",
+    "prettier": "^1.13.5",
```
Comment (amirfefer) [inline package.json:43]: :heart:

**Comment by ohadlevy** [comment]
@sharvit can you rebase please?

**Comment by sharvit** [comment]
Rebased

**Comment by waldenraines** [comment]
So it looks like only https://github.com/theforeman/foreman/pull/6248 remains (and is approved).  Will you rebase this once that PR is merged?

**Comment by sharvit** [comment]
@waldenraines I also created https://github.com/theforeman/foreman/pull/6250

I think I can seperate it out to more micro PRs
- [x] NotificationsDrawer https://github.com/theforeman/foreman/pull/6275
- [x] Statistics https://github.com/theforeman/foreman/pull/6276
- [x] ToastNotifications https://github.com/theforeman/foreman/pull/6277
- [x] Markdown https://github.com/theforeman/foreman/pull/6278
- [x] Diff https://github.com/theforeman/foreman/pull/6274
- [x] Pagination https://github.com/theforeman/foreman/pull/6273

**Comment by sharvit** [comment]
@ohadlevy @waldenraines I think we can merge it now.
Can you rereview please?

**Comment by waldenraines** [comment]
> @ohadlevy @waldenraines I think we can merge it now.
> Can you rereview please?

@sharvit looks like there are codeclimate issues to fix.

**Comment by sharvit** [comment]
@waldenraines - I don't think they are relevant to this PR.
code-climate will track all the lines you changed and will search for issues like functions that are too big, or code that might be reused.

Because this PR is about changing `eslint` configurations and then run the automated script `npm run lint -- --fix`, it changes almost every file but only updates some code styles.
In this case, `code-climate` will find many issues which are not a part of this PR.

File: package.json (hunk 6 above)
Comment (ohadlevy) [reply package.json:15]: @sharvit - bump?

File: webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/BreadcrumbSwitcherToggler.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { Button, Icon } from 'patternfly-react';
 
+// eslint-disable-next-line react/prefer-stateless-function
 class BreadcrumbSwitcherToggler extends React.Component {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/BreadcrumbSwitcherToggler.js:5]: why? can't we convert it to a stateless function?

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/BreadcrumbSwitcherToggler.js:5]: No, since it uses `refs` inside it must be class component.
I actually converted it to a function and then I got an error about it.

See: https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element

File: package.json [hunk 10]
```diff
@@ -11,8 +11,8 @@
     "@storybook/storybook-deployer": "^2.0.0",
     "axios-mock-adapter": "^1.10.0",
     "babel-cli": "^6.10.1",
-    "babel-core": "^6.26.3",
-    "babel-eslint": "^6.1.2",
+    "babel-core": "^6.26.0",
```
Comment (ohadlevy) [inline package.json]: why are you downgrading ?

Comment (sharvit) [reply package.json]: Nice catch! I think it because of some rebase.

File: package.json (hunk 8 above)
Comment (ohadlevy) [reply package.json:43]: @ekohl can you confirm if these changes actually needed to be packaged? afaiu all of these are needed to run linting.

File: package.json (hunk 6 above)
Comment (sharvit) [reply package.json:15]: @ohadlevy are you sure we need a packaging PR here?
I can't find `babel-eslint` in our current packaging.

File: package.json (hunk 10 above)
Comment (sharvit) [reply package.json]: Fixed

**Review by ohadlevy — APPROVED** [review]

**Comment by sharvit** [comment]
Yayy Thanks @ohadlevy

**Comment by glekner** [comment]
🎉 🎉 🎉
